pkg/asset/agent/gencrypto/authconfig.go

```diff
@@ -128,3 +179,65 @@ func localJWTForKey(id string, privateKkeyPem string) (string, error) {
 
 	return tokenString, nil
 }
+
+// InitK8sClient initializes k8s client
+func InitK8sClient(kubeconfig string) (*kubernetes.Clientset, error) {
+	var err error
+	var config *rest.Config
+	if kubeconfig != "" {
+		config, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+	} else {
+		config, err = rest.InClusterConfig()
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	k8sclientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return k8sclientset, err
+}
+
+func (a *AuthConfig) getAuthTokenFromCMElseCreateCMWithAuthToken(k8sclientset kubernetes.Interface, newAgentAuthToken string) (string, error) {
+	cm, err := k8sclientset.CoreV1().ConfigMaps(CMNamespace).Get(context.Background(), CMName, metav1.GetOptions{})
+	if err == nil {
+		// ConfigMap already exists, retrieve the existing agent auth token
+		existingAgentAuthToken, agentAuthTokenExists := cm.Data[CMKey]
```

Comment:
The tokens are non expiring in nature hence the config map does not also expire.

Reply:
I don't think it's acceptable, from a security point of view, to allow (re)using a token that was created for a different session - maybe months before. Similarly to what happens during the installation, every add-nodes session should have its unique token, valid for that session only. A new add-nodes session, that could happen weeks or month later, should use its own token (there could be some particular scenarios to be handled in case of multiple simultaneous create commands, but they look like a pretty edge cases)

Reply:
If we set an expiry time, say 12 hours for example, when the add-nodes command was run for the first time, a secret gets created in the cluster with an expiration time of 12 hours. Now, if a new node is added after 12 hours, then we can update the token value from the secret in the cluster and thereby fulfill the requirement that `every add-nodes session should have its unique token.` However, if another add-nodes is run within 12 hours, we would still be using the old token created during the previous run of the add-nodes command.

I don't see the value of expiring the secret/token. Happy to discuss this further in a 1:1. There is no surety that `every add-nodes session should have its unique token` even with an expiry time set.

Reply:
A token that nevers expires is a a security risk, since it increases the attack surface by allowing a prolonged access in case it was stolen/compromised, whereas short-lived tokens reduce/limit the the window of opportunity for an attacker (ofc this could not be an issue for the installation workflow since it never reuses the same token).
In general all the token based systems offers an [expiration method](https://www.oauth.com/oauth2-servers/access-tokens/access-token-lifetime/), and from a quick check also JWT offers it via the ["exp"](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.4) claim. By looking at the AS code it seems that this field is somehow [managed](https://github.com/openshift/assisted-service/blob/1183f3ac81053b9b7e83db69a8a963da53d494e2/internal/gencrypto/token.go#L107) - at least for the image download service. I'm not sure if it covers natively also other REST api calls, if not it could be directly verified in the installer code by checking the content of the retrieved token `exp` claim field.

Reply:
What should be the token expiry time?

Reply:
I would not exceed 24/48h (in the realistic case I'd expect the nodes to be booted/monitored immediately after having creating the ISO)

Reply:
> so your theory is that people will be more willing to create it on demand?
Right. It should be also easier to re-create an ISO in case of need, given the lower barriers required.

> But it seems to me this could work if you made the token valid for e.g. 48 hours but replaced it with a new one if it is more than 24 hours old.
That seems reasonable.

> How does this work if only the token is stored? Don't we at least need to be able to also retrieve the public key to be able to generate a new node ISO that can validate the existing token? This is not just a bearer token, right?
@pawanpinjarkar could be definitely clarify this point better than me, IIRC it's not required

Reply:
> How does this work if only the token is stored? Don't we at least need to be able to also retrieve the public key to be able to generate a new node ISO that can validate the existing token? This is not just a bearer token, right?

To generate the token we only need the private key. 
Code ref: https://github.com/openshift/installer/blob/b2876fc66fb8292443a3ff19d6db2d9d829e09f2/pkg/asset/agent/gencrypto/authconfig.go#L115-L124

The public key is needed on the assisted service side to verify the signature
Code ref: https://github.com/openshift/assisted-service/blob/644010355ad338b5905fc5f433e2e70de5ae31bf/pkg/auth/agent_local_authenticator.go#L68-L73

It is not a bearer token.